File: rust/src/retrieval/dispatcher.rs
// Copyright (c) 2026 vectorless developers
// SPDX-License-Identifier: Apache-2.0

//! Retrieval dispatcher — the single entry point for all query operations.
//!
//! All queries go through the Orchestrator. There is no separate Worker path.
//! The Orchestrator internally decides whether to run the full analysis phase
//! based on user intent:
//!
//! - **User specified doc_ids** → Orchestrator skips analysis, spawns N Workers
//!   directly (N=1 is a normal case, not special).
//! - **User unspecified (workspace)** → Orchestrator analyzes DocCards, selects
//!   relevant docs, then spawns Workers.
//!
//! Post-processing (synthesis, dedup, rerank) is always unified through the
//! Orchestrator's output — never duplicated in Worker.

use tracing::info;

use crate::agent::config::{AgentConfig, Scope, WorkspaceContext};
use crate::agent::orchestrator::Orchestrator;
use crate::agent::{Agent, EventEmitter, Output};
use crate::error::{Error, Result};
use crate::llm::LlmClient;
use crate::query::QueryPipeline;

/// Dispatch a query to the Orchestrator.
///
/// This is the single entry point from the client layer into the retrieval system.
/// It always goes through the Orchestrator — never directly to Worker.
///
/// Flow:
/// 1. Query understanding via LLM (produces [`QueryPlan`])
/// 2. Orchestrator dispatch (uses QueryPlan for strategy)
///
/// - `Scope::Specified(docs)` → Orchestrator skips analysis, dispatches all docs directly.
/// - `Scope::Workspace(ws)` → Orchestrator runs full flow (analyze → dispatch → fuse → synthesize).
pub async fn dispatch(
    query: &str,
    scope: Scope<'_>,
    config: &AgentConfig,
    llm: &LlmClient,
    emitter: &EventEmitter,
) -> Result<Output> {
    let (ws, skip_analysis) = match scope {
        Scope::Specified(docs) => {
            info!(
                docs = docs.len(),
                "Dispatch (user-specified, skip analysis)"
            );
            (WorkspaceContext::new(docs), true)
        }
        Scope::Workspace(ws) => {
            info!(docs = ws.doc_count(), "Dispatch (workspace, full flow)");
            (ws, false)
        }
    };

    // Step 1: Query understanding — LLM analyzes intent, concepts, complexity.
    // This is required. "Model fails, we fail." — errors propagate.
    info!("Starting query understanding...");
    let query_plan = QueryPipeline::understand(query, llm).await?;

    // Step 2: Dispatch to Orchestrator with the query plan.
    let orchestrator = Orchestrator::new(
        query,
        &ws,
        config.clone(),
        llm.clone(),
        emitter.clone(),
        skip_analysis,
        query_plan,
    );
    orchestrator
        .run()
        .await
        .map_err(|e| Error::Retrieval(e.to_string()))
}

File: vectorless/ask/dispatcher.py
"""Dispatcher — unified entry point for retrieval.

All queries go through dispatch():
1. Query reasoning -> QueryAnalysis (via QueryAnalyzer)
2. Scope resolution -> Specified | Workspace
3. Orchestrator.run() (always)
4. Return Output
"""

from __future__ import annotations

import logging

from vectorless.ask.protocols import DocLoader, EventCallback
from vectorless.ask.errors import AskError
from vectorless.ask.types import DocCard, Output, Specified, Workspace
from vectorless.ask.orchestrator import Orchestrator
from vectorless.ask.reasoning import QueryAnalysis, QueryAnalyzer
from vectorless.llm_client import LLMClient

logger = logging.getLogger(__name__)


async def dispatch(
    query: str,
    scope: Specified | Workspace,
    llm: LLMClient,
    doc_loader: DocLoader,
    event_callback: EventCallback | None = None,
) -> Output:
    """Unified entry point for retrieval.

    - Specified docs -> skip analysis, dispatch Scouts directly
    - Workspace -> analyze (select docs) -> dispatch Scouts -> synthesize
    """
    # Step 1: Query reasoning (multi-stage analysis)
    logger.info("dispatch: query reasoning started")
    analyzer = QueryAnalyzer()
    try:
        query_analysis = await analyzer.analyze(query, llm)
    except AskError:
        raise
    except Exception as e:
        from vectorless.ask.errors import LLMFailureError
        raise LLMFailureError(f"Query analysis failed: {e}") from e
    logger.info(
        "dispatch: query reasoning complete (intent=%s, complexity=%s)",
        query_analysis.intent.value,
        query_analysis.complexity.value,
    )

    # Step 2: Determine skip_analysis from scope
    skip_analysis = isinstance(scope, Specified)
    doc_cards = scope.docs
    logger.info(
        "dispatch: scope=%s docs=%d skip_analysis=%s",
        "specified" if skip_analysis else "workspace",
        len(doc_cards),
        skip_analysis,
    )

    # Step 3: Orchestrator (always)
    logger.info("dispatch: starting orchestrator")
    orchestrator = Orchestrator(
        query=query,
        doc_cards=doc_cards,
        doc_loader=doc_loader,
        llm_client=llm,
        skip_analysis=skip_analysis,
        query_analysis=query_analysis,
        event_callback=event_callback,
    )
    result = await orchestrator.run()
    logger.info("dispatch: orchestrator complete answer_len=%d", len(result.answer) if result.answer else 0)
    return result
